### ISSUE TITLE
Suggestion: simple type-checked pointer primitive

### ISSUE THREAD

**Issue by RReverser** [AssemblyScript/assemblyscript#1363]
In many cases, I find it useful to have type-checked pointers (as opposed to just `usize`) to make sure that values loaded / stored in memory are indeed correct for the given pointer across the FFI boundary.

I've seen https://github.com/AssemblyScript/assemblyscript/issues/1228, and, in particular the Pointer experiment reference there (https://github.com/AssemblyScript/assemblyscript/blob/master/tests/compiler/std/pointer.ts#L3) and I think it's a nice high-level API, but perhaps unnecessarily high-level for majority of cases.

On the other hand, the current primitives (`load` / `store`) are too low-level - they don't provide any typechecking whatsoever, and allow storing arbitrary data to arbitrary pointers.

I propose a solution somewhere in between these two that I've come to use myself - a primitive for opaque pointers:

```ts
@final
//@ts-ignore: decorators
class ptr<T> {
  deref: T;
}
```

This doesn't provide nice high-level APIs for pointer arithmetic, but in many cases it's not necessary. Often enough, as a user, you simply want to accept an opaque typechecked pointer (which is represented as `usize` under the hood), and read or write data to it, and be done with it. This is what such class provides.

`class` is already represented by AssemblyScript as a pointer to the actual data, and in this case data is whatever the `T` is, so you can read and write data via such helper by simply:

```ts
export function doSomethingWithFloat(f: ptr<f32>): void {
  // read float
  let float = f.deref;
  // ...do something
  // write float
  f.deref = float * float;
}
```

Compare this to a "raw" solution:

```ts
export function doSomethingWithFloat2(f: usize): void {
  let float = load<f32>(f);
  // ...do something
  // write float
  store<f32>(f, float * float);
}
```

Both compile to precisely same code (in fact, they get merged by Binaryen if put in the same file), but in the latter case it's too easy for different invocations on the same pointer to accidentally get out of sync in terms of types, as well as public API is less clear on what `f` is supposed to contain.

---

Anyway, just thought I'd post this and would love to hear your thoughts.

**Comment by MaxGraey** [comment]
btw if you want really zero-cost pointer abstraction it better use something like this:
```ts
@final class ptr<T> {
  @inline constructor(offset: usize = 0) {
    return changetype<ptr<T>>(offset);
  }
  @inline get deref(): T {
    return changetype<T>(this);
  }
}
```

**Comment by RReverser** [comment]
The one above is already zero-cost and doesn't require `changetype`, and allows writing as well.

**Comment by MaxGraey** [comment]
It zero-cost only if you store / load from `deref` but when you create it via `new ptr<...>(...)` it alloc memory for  class instance and use also ARC for handling this managed object.

**Comment by RReverser** [comment]
> but when you create it via new ptr<...>(...) it alloc memory for class instance and use also ARC for handling this managed object

I didn't suggest creating pointers - for constructor I was thinking of either making it private (to make it truly opaque and leave up to people to use `changetype`), or providing similar conversion, or just leaving the default as-is in case people actually *want* to create managed pointers.

That's probably worth a separate discussion, my main focus is the `deref` bit.

**Comment by RReverser** [comment]
Btw, your get/set code above doesn't compile in couple of places:

```ts
@final class ptr<T> {
  @inline constructor(offset: usize = 0) {
    return changetype<ptr<T>>(offset);
  }
  @inline get deref(): T {
    return changetype<T>(this); // Type 'main/ptr<f32>' cannot be changed to type 'f32'.
  }
  @inline set deref(value: T): void { // A 'set' accessor cannot have a return type annotation.
    store<T>(changetype<usize>(this), value);
  }
}
```

It's not hard to fix up like this:

```ts
@final class ptr<T> {
  @inline constructor(offset: usize = 0) {
    return changetype<ptr<T>>(offset);
  }
  @inline get deref(): T {
    return load<T>(changetype<usize>(this));
  }
  @inline set deref(value: T) {
    store<T>(changetype<usize>(this), value);
  }
}
```

but just shows yet another reason to avoid bypassing typechecker. After all, static types is the strong suit of TypeScript / AssemblyScript - best to use them directly as much possible, without `changetype` / `load` / `store`.

**Comment by MaxGraey** [comment]
Fixed. This a fiddle: https://webassembly.studio/?f=b1m6z7wfv5h

**Comment by RReverser** [comment]
Yeah I was fixing up in fiddle too.

**Comment by RReverser** [comment]
But, again, compare the complexity of those conversions and just `deref: T;` from my issue - the latter feels much more like a "primitive" and it's properly type-checked without a chance for invalid cast.

What you're designing looks more like the `Pointer` class I referenced in the issue:

> Pointer experiment reference there (https://github.com/AssemblyScript/assemblyscript/blob/master/tests/compiler/std/pointer.ts#L3)

and precisely what I wanted to avoid by suggesting a lower-level primitive.

**Comment by MaxGraey** [comment]
But without that changes `ptr<T>` more look like as `Box<T>` in rust (pointer to heap abstraction) =)

Anyway this suggestion makes sense but in my opinion it should be built-in type for keeping efficiency.

cc @dcodeIO

**Comment by RReverser** [comment]
> pointer to heap abstraction

Only if you create it. If the goal is to just receive them via FFI, it doesn't matter. But if you want to *return* pointers from a function to JS so that it could store and return them later like

```ts
export function foo(): ptr<f32> {
  return new ptr<f32>(1.2);
}
```

then the default heap-based constructor makes sense IMO, since any other pointer wouldn't survive.

**Comment by dcodeIO** [comment]
Perhaps both concepts can be merged into one, so we get the simplicity of the `.deref` field (maybe name it `.value`), plus the utility of the pointer class?

**Comment by RReverser** [comment]
> maybe name it `.value`

I kinda prefer `deref` because it's more explicit about what it is, whereas `value` is a very generic name and `ptr.value` might be ambiguous for a reader (is it value of the pointer as a number or is it dereferenced value?).

> plus the utility of the pointer class?

I think pointer class can extend the low-level primitive, adding more methods, but for those who don't opt-in, they can use `ptr` as simply an opaque pointer. But I don't have strong opinion on this, as long as there is any typechecked pointer type that can be easily used across FFI boundary.

`ptr` as defined in issue description seems very simple to define and well-optimised in self-hosted AssemblyScript yet powerful enough to cover most use-cases already, but if you feel that operator overloads would be commonly used too, I guess they can be added as well.

**Comment by MaxGraey** [comment]
Also if `ptr<T>` will be builtin we could safely call methods which associate to wrapped class like:
```ts
class Foo {
   method(): i32 { ... }
}

let fooPtr = new ptr<Foo>();

fooPtr.method(); // which check holded raw pointer value with null and only after that make `method` call
```
which actually will be:
```ts
let raw = fooPtr.deref;
if (raw !== null) fooPtr.method();
else throw new Error('null deref');
```

**Comment by MaxGraey** [comment]
However perhaps better just make `deref`  as nullable and access to members like:
```ts
let fooPtr = new ptr<Foo>();
fooPtr.deref?.method();
```

**Comment by RReverser** [comment]
@MaxGraey 

1) `ptr<Foo>` is not necessary, since classes are already represented by pointer, so you're basically defining an analogy of `**Foo` from C, not `*Foo`. There are some cases where it's useful, but it won't help with common method calls.

2) Even if this did work, in WebAssembly, 0 adddress is as valid as any other. While C / C++ / Rust target other systems where it's not and so usually restrict pointers to non-null, there's no actual reason to do this in AssemblyScript or any other Wasm-oriented language IMO.

**Comment by MaxGraey** [comment]
Hmm, may be I'm not understand main goals of proposal. I was under impression that ptr<T> is just more safer version of `usize` and mostly needed for handling external interfaces like WASI

**Comment by RReverser** [comment]
> I was under impression that `ptr<T>` is just more safer version of type unsound `usize` and mostly needed for handling external C interfaces like WASI

Yeah, but such wrapping is not necessary for classes, because they are already represented with pointers as-is. It's mostly primitives that are a problem (e.g. `size_t *len` would be represented with `ptr<usize>` not just `usize`).

Well, it will also help with some cases where target expects a `SomeStruct **` (for the function to store its own pointer) - then `ptr<SomeStruct>` would still make sense, and then default heap-based `ptr` constructor is still useful. But that's a less frequent use-case anyway.

**Comment by MaxGraey** [comment]
Oh, it seems you suggest `box<T>` which exactly wrap non-heap primitive types to heap-based managed wrapper.
<img width="245" alt="Снимок экрана 2020-06-28 в 17 08 47" src="https://user-images.githubusercontent.com/1301959/85949898-09c41380-b962-11ea-82c2-838a0af485ea.png">

**Comment by RReverser** [comment]
Not really; `Box` in Rust (assuming that's what you're referring to) is specifically a heap object, while `ptr<T>` in general doesn't make any assumptions about where the data is stored, and can be either on heap or stack or a static variable or anything - it's literally just a generic pointer type.

Just like `T*` in C / C++ can point to any memory location, similarly `ptr<T>` can point to any memory location where the actual `T` is stored.

**Comment by jtenner** [comment]
@RReverser except in Web Assembly the stack is virtual. You can't point to something on the stack. `Box<T>` and `ptr<T>` are equivalent in functionality. At least in web assembly.

**Comment by jtenner** [comment]
I believe I've surfaced complaints about the ergonomics of `changetype<T>()`, and after lots of soul searching and consideration, I've found it to be quite useful.

**Comment by MaxGraey** [comment]
> Not really; Box in Rust (assuming that's what you're referring to) is specifically a heap object, while ptr<T> in general doesn't make any assumptions about where the data is stored, and can be either on heap or stack or a static variable or anything - it's literally just a generic pointer type.

Let's look at another language with managed types, whose boxing was long before Rust. I mean C#. There are has ability to "call-by-value" and "call-by-reference" put "out" or "ref" modifiers before formal and actual parameters like:
```C#
void CallByRefFunc(out int num) {
    num = 111;
}

int res = 0;
CallByRefFunc(out res);
// res has 111 value now
```

The same thing could be simulate via boxing and actually exactly how it works in C# under the hood.

**Comment by RReverser** [comment]
@MaxGraey TBH, I don't know where this discussion is going / what is the purpose, it seems to have become a bit too long and goes into language theory discussions :)

Let's focus on practical side: are you trying to figure out how this wrapper is represented or how it's going to be used or ..? I want to understand how I can improve my explanations above in case they weren't clear.

**Comment by RReverser** [comment]
> except in Web Assembly the stack is virtual. You can't point to something on the stack. `Box<T>` and `ptr<T>` are equivalent in functionality. At least in web assembly.

No, this is not true. In native languages / machines stack and heap also technically live in the same memory, but language-level semantics of `Box` vs general pointers still matter. One is guaranteed to be on the heap, another is a general-purpose address that doesn't care where data is stored.

All in all, I feel that `Box` discussion is derailing from the original purpose and proposal.

**Comment by MaxGraey** [comment]
> Let's focus on practical side: are you trying to figure out how this wrapper is represented or how it's going to be used or ..? I want to understand how I can improve my explanations above in case they weren't clear.

I'm just trying to understand all aspects and find a more generalized solution. Because this proposal looks too specific to support at the language or runtime level.

**Comment by dcodeIO** [comment]
Looks like the confusion here comes from `ptr<T>` being somewhat incomplete, in that it doesn't handle struct-likes in an intuitive way (leading to the `**Foo` situation mentioned above). The [Pointer experiment](https://github.com/AssemblyScript/assemblyscript/blob/master/tests/compiler/std/pointer.ts#L14) tries to tackle that by handling references like structs, since in the reference case, we'd most likely be dealing with an `@unmanaged class` representing the layout of a C-like struct. Hence my suggestion to look into merging both into one to make it a more or less complete concept suitable for most situations.

Regarding Box: Such a pointer and a `Box<T>` are somewhat similar ofc, but the important difference is that `Box<T>` is a managed object (is it?), hence reference counted and whatnot, making it incompatible with `changetype`ing, which is necessary on the boundary. Language-wise I think that an unmanaged pointer interface makes more sense than a box.

**Comment by dcodeIO** [comment]
Regarding making this a complete concept: Perhaps we should go an rename `@unmanaged` to `@struct` to give it proper meaning (have been considering this for a while already), with the following cases supported by `Pointer`:

1. Value: `Pointer<valtype>` is the address of a basic value in memory
2. Class: `Pointer<Class>` is the address of a pointer to a managed class - this is somewhat uncommon
3. Struct: `Pointer<Struct>` is the address of a C-like struct in memory - this is somewhat special

Unfortunately this conflicts with just a `.deref` field because the struct case needs [special inlined logic](https://github.com/AssemblyScript/assemblyscript/blob/2f621a7293aa0f43e8b1e1ca0f768b46f7bc1798/tests/compiler/std/pointer.ts#L14).

**Comment by MaxGraey** [comment]
> Perhaps we should go an rename @unmanaged to @struct

It could lead to wrong assumption about  that class is allocated on stack and passed by value which is not true

**Comment by dcodeIO** [comment]
The renaming is not ultimately necessary if you think that'd be confusing. Making this a complete context would be more important.

**Comment by RReverser** [comment]
@dcodeIO I like your suggestion, although I'm not sure special handling for structs/classes is worth the complexity. To me, it feels a bit magical and I'd rather teach users that objects are already pointers on their own. I'm not strongly opposed to the special handling, just not sure whether it's worth it.

**Comment by RReverser** [comment]
> Regarding Box: Such a pointer and a `Box<T>` are similar ofc, but the important difference is that `Box<T>` is a managed object (is it?), hence reference counted with a GC header and whatnot, making it incompatible with `changetype`ing, which is necessary on the boundary. Language-wise I think that an unmanaged pointer interface makes more sense than a box.

This is a good summary too, by the way. `ptr<T>` is not managed pointer, although it can be constructed from one, of course, just like `Box` / `unique_ptr` from other languages can be turned into a raw pointer.

**Comment by RReverser** [comment]
That seems like a short level for stale issues 😅

**Comment by dcodeIO** [comment]
Whoops, forgot to add a label :)
